### ISSUE TITLE
Docker Service deployed in ECS cluster shows service as running but does not respond

### ISSUE THREAD

**Issue by binitdatta** [aws/amazon-ecs-agent#983]
### Summary

I am developing this (https://github.com/binitdatta/apache-tomcat-legacy-spring-crud-docker-aws-ecs) to explore ECS. This is a legacy Apache Tomcat Spring MVC based war file that accesses an AWS RDS. I can run in locally as well as in a dockerized format in AWS EC2 instance.

Problem starts when I deploy this container to AWS ECS. The service status is RUNNING. However, I cannot access the app at this url : http://52.35.202.115:8080/spring-aws-ecs-demo/students. I also was unable to see the tomcat admin page at http://52.35.202.115:8080

I have debugged the usual suspects like the Security Group port opening. The cluster does have the EC2 instance a member etc.

The Dockerfile and the task definition file is in the repository itself. I used AWS CLI to do the tasks in the following order

### Description

1. 
```
aws ecs create-cluster --cluster-name legacyspringcluster

{
    "cluster": {
        "status": "ACTIVE",
        "clusterName": "legacyspringcluster",
        "registeredContainerInstancesCount": 0,
        "pendingTasksCount": 0,
        "runningTasksCount": 0,
        "activeServicesCount": 0,
        "clusterArn": "arn:aws:ecs:us-west-2:107222267664:cluster/legacyspringcluster"
    }
}
```

2. 
```
aws ecs describe-clusters --clusters legacyspringcluster

{
    "clusters": [
        {
            "status": "ACTIVE",
            "clusterName": "legacyspringcluster",
            "registeredContainerInstancesCount": 0,
            "pendingTasksCount": 0,
            "runningTasksCount": 0,
            "activeServicesCount": 0,
            "clusterArn": "arn:aws:ecs:us-west-2:107222267664:cluster/legacyspringcluster"
        }
    ],
    "failures": []
}
```

3. Create a file ecs.config.sh
```
#!/bin/bash
echo ECS_CLUSTER=javalegacyspringwarcluster >> /etc/ecs/ecs.config

chmod 777 ecs.config.sh
```

4. Create an EC2 ECS Cluster Member Instance
```
aws ec2 run-instances --image-id ami-62d35c02 --count 1 --instance-type t2.micro --iam-instance-profile Name=ecsInstanceRole --key-name BinitwmpAccessKey --security-group-ids sg-ef6abe89 --subnet subnet-13752964 --associate-public-ip-address --user-data file://ecs.config.sh
```
5. 
```
aws ec2 create-tags --resources i-0e64693874a46ea77 --tags Key=Name,Value=CLI_BinitJavaLegacyWarWMPECSInstance
```
6. `aws ec2 describe-instance-status --instance-id i-0d78b35e36faa2476`

7. 
```
aws ecs list-container-instances --cluster legacyspringcluster

{
    "containerInstanceArns": [
        "arn:aws:ecs:us-west-2:107222267664:container-instance/b10db456-7a04-4fd5-bfe9-2029beffbd2c"
    ]
}
```

8. 
```
aws ecr create-repository --repository-name binitdatta/spring-aws-ecs-demo

{
    "repository": {
        "registryId": "107222267664",
        "repositoryName": "binitdatta/spring-aws-ecs-demo",
        "repositoryArn": "arn:aws:ecr:us-west-2:107222267664:repository/binitdatta/spring-aws-ecs-demo",
        "createdAt": 1505327080.0,
        "repositoryUri": "107222267664.dkr.ecr.us-west-2.amazonaws.com/binitdatta/spring-aws-ecs-demo"
    }
}
```
9. Authenticate your Docker client to Amazon ECR
```
aws ecr get-login --no-include-email
```
10. Tag the spring-aws-ecs-demo image
```
docker tag spring-aws-ecs-demo 107222267664.dkr.ecr.us-west-2.amazonaws.com/binitdatta/spring-aws-ecs-demo:latest

docker push 107222267664.dkr.ecr.us-west-2.amazonaws.com/binitdatta/spring-aws-ecs-demo:latest
```

11. Create a new file spring-aws-ecs-demo.json
```
{
  "containerDefinitions": [
    {
      "name": "spring-aws-ecs-demo",
      "image": "107222267664.dkr.ecr.us-west-2.amazonaws.com/binitdatta/spring-aws-ecs-demo:latest",
      "portMappings": [
        {
          "containerPort": 8080,
          "hostPort": 8080
        }
      ],
      "memory": 100,
      "cpu": 102
    }
  ],
  "family": "spring-aws-ecs-demo"
}
```

12. 
```
aws ecs register-task-definition --cli-input-json file://spring-aws-ecs-demo.json

{
    "taskDefinition": {
        "status": "ACTIVE",
        "family": "spring-aws-ecs-demo",
        "placementConstraints": [],
        "requiresAttributes": [
            {
                "name": "com.amazonaws.ecs.capability.ecr-auth"
            }
        ],
        "volumes": [],
        "taskDefinitionArn": "arn:aws:ecs:us-west-2:107222267664:task-definition/spring-aws-ecs-demo:3",
        "containerDefinitions": [
            {
                "environment": [],
                "name": "spring-aws-ecs-demo",
                "mountPoints": [],
                "image": "107222267664.dkr.ecr.us-west-2.amazonaws.com/binitdatta/spring-aws-ecs-demo:latest",
                "cpu": 102,
                "portMappings": [
                    {
                        "protocol": "tcp",
                        "containerPort": 8080,
                        "hostPort": 8080
                    }
                ],
                "memory": 100,
                "essential": true,
                "volumesFrom": []
            }
        ],
        "revision": 3
    }
}
```

13. 
```
aws ecs create-service --cluster legacyspringcluster --service-name spring-aws-ecs-demo --task-definition spring-aws-ecs-demo --desired-count 1

{
    "service": {
        "status": "ACTIVE",
        "taskDefinition": "arn:aws:ecs:us-west-2:107222267664:task-definition/spring-aws-ecs-demo:3",
        "pendingCount": 0,
        "loadBalancers": [],
        "placementConstraints": [],
        "createdAt": 1505327874.657,
        "desiredCount": 1,
        "serviceName": "spring-aws-ecs-demo",
        "clusterArn": "arn:aws:ecs:us-west-2:107222267664:cluster/legacyspringcluster",
        "serviceArn": "arn:aws:ecs:us-west-2:107222267664:service/spring-aws-ecs-demo",
        "deploymentConfiguration": {
            "maximumPercent": 200,
            "minimumHealthyPercent": 100
        },
        "deployments": [
            {
                "status": "PRIMARY",
                "pendingCount": 0,
                "createdAt": 1505327874.657,
                "desiredCount": 1,
                "taskDefinition": "arn:aws:ecs:us-west-2:107222267664:task-definition/spring-aws-ecs-demo:3",
                "updatedAt": 1505327874.657,
                "id": "ecs-svc/9223370531526901150",
                "runningCount": 0
            }
        ],
        "events": [],
        "runningCount": 0,
        "placementStrategy": []
    }
}
```

14. Wait till AWS ECs Console shows RUNNING for Service and desired count is same as the expected count.

15. Access : `http://52.35.202.115:8080/spring-aws-ecs-demo/students`

16. Result Takes a long time and then gives blank page.

17. Docker ps output on the ECS Cluster member EC2 instance
```
CONTAINER ID        IMAGE                                                                                COMMAND             CREATED             STATUS              PORTS                    NAMES
25c1aec5d147        107222267664.dkr.ecr.us-west-2.amazonaws.com/binitdatta/spring-aws-ecs-demo:latest   "catalina.sh run"   8 minutes ago       Up 8 minutes        0.0.0.0:8080->8080/tcp   ecs-spring-aws-ecs-demo-3-spring-aws-ecs-demo-e8cb86e899edcdb96d00
2872df9538d8        amazon/amazon-ecs-agent:latest                                                       "/agent"            34 minutes ago      Up 34 minutes                                ecs-agent
```

18. Questions:

1. How do I see logs of my service. docker logs 25c1aec5d147 shows half the logs

2. The service status shows RUNNING but http://52.35.202.115:8080/spring-aws-ecs-demo/students is not accessible. 

Please help,

Thanks,

Binit

### Expected Behavior

This should bring up the Web UI

http://52.35.202.115:8080/spring-aws-ecs-demo/students

### Observed Behavior

http://52.35.202.115:8080/spring-aws-ecs-demo/students is giving blank page

**Comment by aaithal** [comment]
@binitdatta  thank you so much for the detailed description of the issue. 

> How do I see logs of my service. 

You can use the `docker logs ` command to view the container logs. If the container is additionally generating log files, you can attach to the running container and inspect the file system for logs as well. 

> The service status shows RUNNING but http://52.35.202.115:8080/spring-aws-ecs-demo/students is not accessible.

Since the container is running successfully, I'd venture a guess that this is more of an application/config issue that you're running into. Can you try the following and see if it's helpful in resolving the issue?

1. Does `curl localhost:8080/spring-aws-ecs-demo/students` return success? If that's the case, may be there's a security group issue

2. Can you try running this docker container without ECS and see if you can access the endpoint at port 8080? You'll have to basically drop the service's desired count to zero and invoke the `docker run` command (something to the effect of `docker run -P 107222267664.dkr.ecr.us-west-2.amazonaws.com/binitdatta/spring-aws-ecs-demo:latest` with both `--net bridge` and `--net host` . You can then check if you're able to reach the container that's launched using those commands.

Thanks,
Anirudh

**Comment by binitdatta** [comment]
@aaithal ,

Thanks for your response. I very much appreciate your help. 

1. I have already ran the application in Docker within AWS EC2 instance and it was all good. I am using the same Security Group here for the ECS instance as well. 

2. I did try docker logs <container_id> and it shows only a portion of the log file. Will try again and post the log file and the portion not shown.

3. I understand that ECS uses two IAM roles. One is the Instance role and another is the service role. The service role is probably used for accessing ELBs. As I do not have ELB in this POC, I only used the instance role. Not sure if this could be an issue.

Will try again with your suggestions and post by a day or two,

Thanks again for helping.,

Thanks,
Binit
